src/libstore/build/derivation-goal.cc

```diff
@@ -143,6 +143,115 @@ void DerivationGoal::work()
     (this->*state)();
 }
 
+static std::regex shVarName("[A-Za-z_][A-Za-z0-9_]*");
+std::optional<StructuredAttrsWithShellRC> DerivationGoal::generateStructuredAttrs(
+    std::optional<StringMap> inputRewrites)
+{
+    if (!parsedDrv) return std::nullopt;
+    auto structuredAttrs = parsedDrv->getStructuredAttrs();
+    if (!structuredAttrs) return std::nullopt;
+
+    auto json = *structuredAttrs;
+
+    /* Add an "outputs" object containing the output paths. */
+    nlohmann::json outputs;
+    for (auto & i : drv->outputs) {
+        if (inputRewrites) {
+            /* The placeholder must have a rewrite, so we use it to cover both the
+               cases where we know or don't know the output path ahead of time. */
+            outputs[i.first] = rewriteStrings(hashPlaceholder(i.first), inputRewrites.value());
+        } else {
+            /* This case is only relevant for the nix-shell */
+            outputs[i.first] = hashPlaceholder(i.first);
+        }
+    }
+    json["outputs"] = outputs;
+
+    /* Handle exportReferencesGraph. */
+    auto e = json.find("exportReferencesGraph");
+    if (e != json.end() && e->is_object()) {
+        for (auto i = e->begin(); i != e->end(); ++i) {
```

Comment:
AFAIU this won't work in both cases in the structured attrs stuff as we need both `*i` (i.e. the value) and `i.key()`.

Reply:
Oh yes, then you can do something like 

```c++
for (auto & [key, value] : e)
  …
```

(There’s an example of that in somewhere in the big code snippet under https://github.com/nlohmann/json#stl-like-access).

But it’s not like it’s new code either, so it can probably also be left for later